lib/core/test/unit/Cardano/Wallet/Primitive/CollateralSpec.hs

```diff
@@ -0,0 +1,774 @@
+
+module Cardano.Wallet.Primitive.CollateralSpec where
+
+import Prelude
+
+import Cardano.Wallet.Primitive.Collateral
+    ( AddrNotSuitableForCollateral (..)
+    , AddressType (..)
+    , Credential (..)
+    , addressType
+    , addressTypeFromHeaderNibble
+    , addressTypeSuitableForCollateral
+    , addressTypeToHeaderNibble
+    , asCollateral
+    , classifyCollateralAddress
+    , getAddressType
+    , putAddressType
+    )
+import Cardano.Wallet.Primitive.Types.Address
+    ( Address (..) )
+import Cardano.Wallet.Primitive.Types.TokenBundle
+    ( TokenBundle )
+import Cardano.Wallet.Primitive.Types.TokenBundle.Gen
+    ( genTokenBundleSmallRangePositive, shrinkTokenBundleSmallRangePositive )
+import Cardano.Wallet.Primitive.Types.Tx
+    ( TxOut (..) )
+import Cardano.Wallet.Unsafe
+    ( unsafeBech32Decode )
+import Control.Monad
+    ( replicateM_ )
+import Data.ByteString
+    ( ByteString )
+import Data.ByteString.Base58
+    ( bitcoinAlphabet, decodeBase58 )
+import Data.Function
+    ( (&) )
+import Data.Maybe
+    ( fromJust, isJust, isNothing )
+import Numeric
+    ( showHex )
+import Test.Hspec
+    ( Expectation, Spec, describe, it, parallel, shouldBe )
+import Test.QuickCheck
+    ( Arbitrary (..)
+    , Gen
+    , Property
+    , checkCoverage
+    , counterexample
+    , cover
+    , coverTable
+    , disjoin
+    , forAll
+    , forAllShrink
+    , frequency
+    , oneof
+    , property
+    , tabulate
+    , withMaxSuccess
+    , (===)
+    )
+import Test.QuickCheck.Hedgehog
+    ( hedgehog )
+
+import qualified Cardano.Ledger.Address as L
+import qualified Cardano.Ledger.Credential as L
+import qualified Cardano.Ledger.Crypto as CC
+import qualified Cardano.Ledger.Hashes as L
+import qualified Cardano.Wallet.Primitive.Types.TokenBundle as TokenBundle
+import qualified Data.Binary.Get as B
+import qualified Data.Binary.Put as B
+import qualified Data.ByteString as BS
+import qualified Data.ByteString.Lazy as BL
+import qualified Test.Cardano.Chain.Common.Gen as Byron
+import qualified Test.Shelley.Spec.Ledger.Serialisation.EraIndepGenerators as L
+import qualified Test.Shelley.Spec.Ledger.Serialisation.Generators.Genesis as L
+    ( genRewardAcnt )
+
+-- To begin with, we will write our generators and tests for the @AddressType@
+-- type.
+--
+-- First we write our generator, which customizes its frequency so that every
+-- branch of the AddressType type is evenly covered:
+
+-- | Generate an AddressType.
+genAddressType :: Gen AddressType
+genAddressType =
+    frequency
+        [ (4, BaseAddress <$> genCredential <*> genCredential)
+        , (2, PointerAddress <$> genCredential)
+        , (2, EnterpriseAddress <$> genCredential)
+        , (2, StakeAddress <$> genCredential)
+        , (1, pure BootstrapAddress)
+        ]
+
+-- | Generate a credential.
+genCredential :: Gen Credential
+genCredential =
+    oneof [ pure CredentialKeyHash
+          , pure CredentialScriptHash
+          ]
```

Comment:
Use 4-char indents rather than alignment here: (https://github.com/input-output-hk/adrestia/wiki/Coding-Standards#avoid-variable-length-indentation)

```suggestion
genCredential :: Gen Credential
genCredential = oneof
    [ pure CredentialKeyHash
    , pure CredentialScriptHash
    ]
```